sql/hive/src/test/scala/org/apache/spark/sql/hive/ParquetEncryptionSuite.scala

```diff
@@ -0,0 +1,148 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.hive
+
+import java.io.File
+import java.nio.charset.StandardCharsets
+import java.util.{Base64, HashMap, Map}
+
+import scala.sys.process._
+
+import org.apache.hadoop.conf.Configuration
+import org.apache.parquet.crypto.{KeyAccessDeniedException, ParquetCryptoRuntimeException}
+import org.apache.parquet.crypto.keytools.{KeyToolkit, KmsClient}
+
+import org.apache.spark.sql.QueryTest
+import org.apache.spark.sql.test.SharedSparkSession
+
+/**
+ * A test suite that tests parquet modular encryption usage.
+ */
+class ParquetEncryptionSuite extends QueryTest with SharedSparkSession {
+
+  private val encoder = Base64.getEncoder
+  private val footerKey =
+    encoder.encodeToString("0123456789012345".getBytes(StandardCharsets.UTF_8))
+  private val key1 = encoder.encodeToString("1234567890123450".getBytes(StandardCharsets.UTF_8))
+  private val key2 = encoder.encodeToString("1234567890123451".getBytes(StandardCharsets.UTF_8))
+
+  import testImplicits._
+
+  test("SPARK-34990: Write and read an encrypted parquet") {
+    withTempDir { dir =>
+      withSQLConf(
+        "parquet.crypto.factory.class" ->
+          "org.apache.parquet.crypto.keytools.PropertiesDrivenCryptoFactory",
+        "parquet.encryption.kms.client.class" -> "org.apache.spark.sql.hive.InMemoryKMS",
+        "parquet.encryption.key.list" ->
+          s"footerKey: ${footerKey}, key1: ${key1}, key2: ${key2}") {
+
+        val inputDF = Seq((1, 22, 333)).toDF("a", "b", "c")
+        val parquetDir = new File(dir, "parquet").getCanonicalPath
+        inputDF.write
+          .option("parquet.encryption.column.keys", "key1: a, b; key2: c")
+          .option("parquet.encryption.footer.key", "footerKey")
+          .parquet(parquetDir)
+
+        verifyParquetEncrypted(parquetDir)
+
+        val parquetDF = spark.read.parquet(parquetDir)
+        assert(parquetDF.inputFiles.nonEmpty)
+        val readDataset = parquetDF.select("a", "b", "c")
+        checkAnswer(readDataset, inputDF)
+      }
+    }
+  }
+
+  private def verifyParquetEncrypted(parquetDir: String) = {
+    val parquetPartitionFile = Seq("ls", "-tr", parquetDir).!!.split("\\s+")(0)
+    val fullFilename = parquetDir + "/" + parquetPartitionFile
+    val magic = Seq("tail", "-c", "4", fullFilename).!!
+    assert(magic.stripLineEnd.trim() == "PARE")
+  }
+}
+
+/**
+ * This is a mock class, built just for parquet encryption testing in Spark
+ * and based on InMemoryKMS in parquet-hadoop tests.
+ * Don't use it as an example of a KmsClient implementation.
+ * Use parquet-hadoop/src/test/java/org/apache/parquet/crypto/keytools/samples/VaultClient.java
+ * as a sample implementation instead.
+ */
+class InMemoryKMS extends KmsClient {
+  private var masterKeyMap: Map[String, Array[Byte]] = null
```

Comment:
What is the difference between this InMemoryKMS and the one in implemented in parquet-hadoop tests? 
Could we use that one by adding parquet-hadoop test as a maven dependency?

They are publishing the jar, check https://repo1.maven.org/maven2/org/apache/parquet/parquet-hadoop/1.12.0/ :
`parquet-hadoop-1.12.0-tests.jar                   2021-03-17 08:57    411020`